onnx/shape_inference.py

```diff
@@ -32,68 +34,36 @@ def infer_shapes(
     bug in shape inference), and the result is unspecified.
 
     Arguments:
-        model: ModelProto.
+        model: ModelProto. If the model bytes size is larger than 2GB, function should be called using model path.
         check_type: Checks the type-equality for input and output.
-        strict_mode: Stricter shape inference, it will throw errors if any;
-            Otherwise, simply stop if any error.
+        strict_mode: Stricter shape inference, it will throw errors if any; Otherwise, simply stop if any error.
         data_prop: Enables data propagation for limited operators to perform shape computation.
+        output_path: Must provide `output_path` if `model` is a path.
 
     Returns:
-        (ModelProto) model with inferred shape information
+        (ModelProto) model with inferred shape information. Return None if `model` is a path.
     """
-    if isinstance(model, (ModelProto, bytes)):
-        model_str = model if isinstance(model, bytes) else model.SerializeToString()
+    # If model is a path instead of ModelProto
+    if isinstance(model, (str, os.PathLike)):
+        model_path = os.fspath(model)
+        if output_path == "":
+            raise ValueError("output_path must be provided when input is a path.")
+        output_path = os.fspath(output_path)
+        C.infer_shapes_path(model_path, output_path, check_type, strict_mode, data_prop)
+    else:
+        protobuf_string = (
+            model if isinstance(model, bytes) else model.SerializeToString()
+        )
+        # If the protobuf is larger than 2GB,
+        # remind users should use the model path to check
+        if sys.getsizeof(protobuf_string) > onnx.checker.MAXIMUM_PROTOBUF:
+            raise ValueError(
+                "This protobuf of onnx model is too large (>2GB). Call infer_shapes with model path instead."
+            )
         inferred_model_str = C.infer_shapes(
-            model_str, check_type, strict_mode, data_prop
+            protobuf_string, check_type, strict_mode, data_prop
         )
         return onnx.load_from_string(inferred_model_str)
-    if isinstance(model, str):
-        raise TypeError(
-            "infer_shapes only accepts ModelProto or bytes,"
-            "you can use infer_shapes_path for the model path (String)."
-        )
-
-    raise TypeError(
-        f"infer_shapes only accepts ModelProto or bytes, incorrect type: {type(model)}"
-    )
-
-
-def infer_shapes_path(
```

Comment:
It is part of the public API. It should remain marked as deprecated for one release.

Reply:
Thank you, I put it back. May I know how to mark it as deprecated?

Reply:
There is no common function to do that right now. You can add one or raise a DeprecationWarning saying it will be removed in version x.xx.

Reply:
Thank you. I added `DeprecationWarning` messages.